File: tools/tf-types-ts/src/core/session-migration.ts
/**
 * Session migration helpers (TF-0003).
 *
 * `migrateSession` produces a signed `SessionMigration` record describing
 * a session moving between transports while preserving session_id and
 * generation. The session itself stays alive — only its transport
 * binding changes — so authority grants, capability tokens, and proof
 * lineage carry over.
 *
 * `verifySessionMigration` re-derives the digest and verifies the
 * ed25519 signature, plus enforces monotonic generation.
 *
 * `Ratchet` implements TF-0003's double-ratchet auto-trigger: the
 * caller signals that N messages or T seconds have elapsed and the
 * ratchet rotates symmetric keys via HKDF-SHA256(prev_key,
 * info="tf-session/ratchet"). It does not bind to the actual session
 * cipher — that's the SessionEndpoint's job — but it provides a
 * deterministic key sequence so both peers stay in lock-step.
 */

import type {
  ActorId,
  Capability,
  Timestamp,
  SignatureEnvelope,
} from "../generated/_common.js";
import type { TransportBinding } from "../generated/transport-binding.js";
import type { SessionMigration } from "../generated/session-migration.js";
import { canonicalize } from "./canonical.js";
import { ed25519Sign, ed25519Verify, hkdfSha256, sha256 } from "./crypto.js";

export interface MigrateSessionArgs {
  sessionId: string;
  generation: number;
  fromBinding: TransportBinding;
  toBinding: TransportBinding;
  rotatedKeys?: boolean;
  preservedCapabilities?: Capability[];
  reason?: string;
  signer: ActorId;
  privateKey: Uint8Array;
  migratedAt?: Timestamp;
}

export function migrationSigningBytes(m: SessionMigration): Uint8Array {
  const { signature: _signature, ...rest } = m;
  void _signature;
  return sha256(new TextEncoder().encode(canonicalize(rest as unknown)));
}

export async function migrateSession(args: MigrateSessionArgs): Promise<SessionMigration> {
  if (args.generation < 1) {
    throw new Error("session migration generation must be >= 1");
  }
  const migratedAt = args.migratedAt ?? new Date().toISOString();
  const draft: SessionMigration = {
    migration_version: "1",
    session_id: args.sessionId,
    generation: args.generation,
    from_binding: args.fromBinding,
    to_binding: args.toBinding,
    migrated_at: migratedAt,
    signer: args.signer,
    signature: { algorithm: "ed25519", signer: args.signer, signature: "" } as SignatureEnvelope,
  };
  if (args.rotatedKeys) draft.rotated_keys = true;
  if (args.preservedCapabilities && args.preservedCapabilities.length > 0) {
    draft.preserved_capabilities = args.preservedCapabilities;
  }
  if (args.reason) draft.reason = args.reason;
  const digest = migrationSigningBytes(draft);
  const sig = await ed25519Sign(digest, args.privateKey);
  draft.signature = {
    algorithm: "ed25519",
    signer: args.signer,
    signature: Buffer.from(sig).toString("base64"),
  };
  return draft;
}

export interface VerifySessionMigrationArgs {
  migration: SessionMigration;
  publicKey: Uint8Array;
  /** Last generation observed by the verifier; if set, the new
   *  generation must be strictly greater. */
  lastGeneration?: number;
  /** Expected session_id; if set, must match exactly. */
  expectedSessionId?: string;
}

export interface VerifySessionMigrationResult {
  ok: boolean;
  reason?: string;
}

export async function verifySessionMigration(
  args: VerifySessionMigrationArgs,
): Promise<VerifySessionMigrationResult> {
  const m = args.migration;
  if (m.migration_version !== "1") {
    return { ok: false, reason: `unsupported migration_version ${m.migration_version}` };
  }
  if (m.signature.signer !== m.signer) {
    return { ok: false, reason: "signature signer does not match signer" };
  }
  if (m.signature.algorithm !== "ed25519") {
    return { ok: false, reason: `unsupported signature algorithm ${m.signature.algorithm}` };
  }
  if (args.expectedSessionId && m.session_id !== args.expectedSessionId) {
    return { ok: false, reason: "session_id mismatch" };
  }
  if (args.lastGeneration !== undefined && m.generation <= args.lastGeneration) {
    return { ok: false, reason: `generation ${m.generation} <= last seen ${args.lastGeneration} (replay)` };
  }
  const digest = migrationSigningBytes(m);
  const sigBytes = new Uint8Array(Buffer.from(m.signature.signature, "base64"));
  const verified = await ed25519Verify(args.publicKey, digest, sigBytes);
  if (!verified) {
    return { ok: false, reason: "migration signature did not verify" };
  }
  return { ok: true };
}

/* -------------------------------------------------------------------------- */
/*  Double ratchet                                                            */
/* -------------------------------------------------------------------------- */

const RATCHET_INFO = new TextEncoder().encode("tf-session/ratchet");

export interface RatchetOptions {
  /** Rotate after this many messages have been processed. Default 1024. */
  maxMessages?: number;
  /** Rotate after this many seconds have elapsed since the last rotation.
   *  Default 600 seconds. */
  maxAgeSeconds?: number;
}

export class Ratchet {
  private readonly maxMessages: number;
  private readonly maxAgeMs: number;
  private currentKey: Uint8Array;
  private rotationCount = 0;
  private messagesSinceRotation = 0;
  private lastRotationMs: number;

  constructor(initialKey: Uint8Array, opts: RatchetOptions = {}) {
    if (initialKey.length !== 32) {
      throw new Error("ratchet key must be 32 bytes");
    }
    this.currentKey = new Uint8Array(initialKey);
    this.maxMessages = opts.maxMessages ?? 1024;
    this.maxAgeMs = (opts.maxAgeSeconds ?? 600) * 1000;
    this.lastRotationMs = Date.now();
  }

  /** Current 32-byte symmetric key. */
  key(): Uint8Array {
    return new Uint8Array(this.currentKey);
  }

  /** How many rotations have happened since construction. */
  generation(): number {
    return this.rotationCount;
  }

  /** Should be called on every successful frame send/receive. Returns
   *  true if the ratchet rotated as a result. */
  observeMessage(): boolean {
    this.messagesSinceRotation += 1;
    if (this.shouldRotate()) {
      this.rotate();
      return true;
    }
    return false;
  }

  /** Force a rotation right now. */
  rotate(): void {
    this.currentKey = hkdfSha256(this.currentKey, new Uint8Array(0), RATCHET_INFO, 32);
    this.rotationCount += 1;
    this.messagesSinceRotation = 0;
    this.lastRotationMs = Date.now();
  }

  private shouldRotate(): boolean {
    if (this.messagesSinceRotation >= this.maxMessages) return true;
    if (Date.now() - this.lastRotationMs >= this.maxAgeMs) return true;
    return false;
  }
}

File: crates/tf-types/src/session_migration.rs
//! Session-migration helpers (Rust mirror of TS).

use base64::engine::general_purpose::STANDARD;
use base64::Engine;
use ed25519_dalek::{Signature, Signer, SigningKey, Verifier, VerifyingKey};
use hkdf::Hkdf;
use serde::{Deserialize, Serialize};
use serde_json::Value;
use sha2::{Digest, Sha256};

use crate::canonicalize;

#[derive(Clone, Debug, Serialize, Deserialize, PartialEq, Eq)]
pub struct TransportBinding {
    pub binding_version: String,
    pub kind: String,
    #[serde(skip_serializing_if = "Option::is_none", default)]
    pub endpoint: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none", default)]
    pub exporter_key: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none", default)]
    pub peer_cert_fingerprint: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none", default)]
    pub tls_alpn: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none", default)]
    pub established_at: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none", default)]
    pub metadata: Option<Value>,
}

#[derive(Clone, Debug, Serialize, Deserialize, PartialEq, Eq)]
pub struct SessionMigration {
    pub migration_version: String,
    pub session_id: String,
    pub generation: u64,
    pub from_binding: TransportBinding,
    pub to_binding: TransportBinding,
    #[serde(skip_serializing_if = "Option::is_none", default)]
    pub preserved_capabilities: Option<Vec<Value>>,
    #[serde(skip_serializing_if = "Option::is_none", default)]
    pub rotated_keys: Option<bool>,
    pub migrated_at: String,
    #[serde(skip_serializing_if = "Option::is_none", default)]
    pub reason: Option<String>,
    pub signer: String,
    pub signature: SignatureEnvelope,
}

#[derive(Clone, Debug, Serialize, Deserialize, PartialEq, Eq)]
pub struct SignatureEnvelope {
    pub algorithm: String,
    pub signer: String,
    pub signature: String,
}

pub fn migration_signing_bytes(m: &SessionMigration) -> [u8; 32] {
    let mut value = serde_json::to_value(m).unwrap_or(Value::Null);
    if let Value::Object(map) = &mut value {
        map.remove("signature");
    }
    let canonical = canonicalize(&value).unwrap_or_default();
    Sha256::digest(canonical.as_bytes()).into()
}

#[allow(clippy::too_many_arguments)]
pub fn migrate_session(
    session_id: &str,
    generation: u64,
    from_binding: TransportBinding,
    to_binding: TransportBinding,
    rotated_keys: bool,
    reason: Option<&str>,
    signer: &str,
    private_key: &[u8; 32],
    migrated_at: Option<&str>,
) -> SessionMigration {
    let migrated_at = migrated_at.map(str::to_string).unwrap_or_else(now_iso8601);
    let mut m = SessionMigration {
        migration_version: "1".into(),
        session_id: session_id.into(),
        generation,
        from_binding,
        to_binding,
        preserved_capabilities: None,
        rotated_keys: if rotated_keys { Some(true) } else { None },
        migrated_at,
        reason: reason.map(str::to_string),
        signer: signer.into(),
        signature: SignatureEnvelope {
            algorithm: "ed25519".into(),
            signer: signer.into(),
            signature: String::new(),
        },
    };
    let digest = migration_signing_bytes(&m);
    let signing = SigningKey::from_bytes(private_key);
    let sig: Signature = signing.sign(&digest);
    m.signature.signature = STANDARD.encode(sig.to_bytes());
    m
}

#[derive(Debug)]
pub struct VerifyMigrationResult {
    pub ok: bool,
    pub reason: Option<String>,
}

pub fn verify_session_migration(
    m: &SessionMigration,
    public_key: &[u8; 32],
    last_generation: Option<u64>,
    expected_session_id: Option<&str>,
) -> VerifyMigrationResult {
    let rejected = |r: &str| VerifyMigrationResult {
        ok: false,
        reason: Some(r.to_string()),
    };
    if m.migration_version != "1" {
        return rejected(&format!(
            "unsupported migration_version {}",
            m.migration_version
        ));
    }
    if m.signature.signer != m.signer {
        return rejected("signature signer does not match signer");
    }
    if m.signature.algorithm != "ed25519" {
        return rejected(&format!(
            "unsupported signature algorithm {}",
            m.signature.algorithm
        ));
    }
    if let Some(expected) = expected_session_id {
        if m.session_id != expected {
            return rejected("session_id mismatch");
        }
    }
    if let Some(last) = last_generation {
        if m.generation <= last {
            return rejected(&format!(
                "generation {} <= last seen {} (replay)",
                m.generation, last
            ));
        }
    }
    let digest = migration_signing_bytes(m);
    let sig_bytes = match STANDARD.decode(&m.signature.signature) {
        Ok(b) => b,
        Err(e) => return rejected(&format!("signature base64 decode: {}", e)),
    };
    let sig = match Signature::from_slice(&sig_bytes) {
        Ok(s) => s,
        Err(e) => return rejected(&format!("signature parse: {}", e)),
    };
    let vk = match VerifyingKey::from_bytes(public_key) {
        Ok(v) => v,
        Err(e) => return rejected(&format!("verifying key: {}", e)),
    };
    if vk.verify(&digest, &sig).is_err() {
        return rejected("migration signature did not verify");
    }
    VerifyMigrationResult {
        ok: true,
        reason: None,
    }
}

const RATCHET_INFO: &[u8] = b"tf-session/ratchet";

#[derive(Debug)]
pub struct Ratchet {
    current_key: [u8; 32],
    rotation_count: u64,
    messages_since_rotation: u32,
    max_messages: u32,
}

impl Ratchet {
    pub fn new(initial_key: [u8; 32], max_messages: Option<u32>) -> Self {
        Ratchet {
            current_key: initial_key,
            rotation_count: 0,
            messages_since_rotation: 0,
            max_messages: max_messages.unwrap_or(1024),
        }
    }

    pub fn key(&self) -> [u8; 32] {
        self.current_key
    }

    pub fn generation(&self) -> u64 {
        self.rotation_count
    }

    pub fn observe_message(&mut self) -> bool {
        self.messages_since_rotation += 1;
        if self.messages_since_rotation >= self.max_messages {
            self.rotate();
            true
        } else {
            false
        }
    }

    pub fn rotate(&mut self) {
        let hk = Hkdf::<Sha256>::new(None, &self.current_key);
        let mut next = [0u8; 32];
        hk.expand(RATCHET_INFO, &mut next).expect("hkdf");
        self.current_key = next;
        self.rotation_count += 1;
        self.messages_since_rotation = 0;
    }
}

fn now_iso8601() -> String {
    let secs = std::time::SystemTime::now()
        .duration_since(std::time::UNIX_EPOCH)
        .unwrap_or_default()
        .as_secs() as i64;
    let (y, m, d, h, mi, s) = secs_to_ymdhms(secs);
    format!("{:04}-{:02}-{:02}T{:02}:{:02}:{:02}Z", y, m, d, h, mi, s)
}

fn secs_to_ymdhms(secs: i64) -> (i32, u32, u32, u32, u32, u32) {
    let days = secs.div_euclid(86_400);
    let time = secs.rem_euclid(86_400);
    let hour = (time / 3600) as u32;
    let minute = ((time % 3600) / 60) as u32;
    let second = (time % 60) as u32;
    let z = days + 719_468;
    let era = if z >= 0 { z } else { z - 146_096 } / 146_097;
    let doe = (z - era * 146_097) as u64;
    let yoe = (doe - doe / 1460 + doe / 36524 - doe / 146096) / 365;
    let y = yoe as i64 + era * 400;
    let doy = doe - (365 * yoe + yoe / 4 - yoe / 100);
    let mp = (5 * doy + 2) / 153;
    let d = (doy - (153 * mp + 2) / 5 + 1) as u32;
    let m = if mp < 10 {
        (mp + 3) as u32
    } else {
        (mp - 9) as u32
    };
    let year = if m <= 2 { y + 1 } else { y };
    (year as i32, m, d, hour, minute, second)
}
